### ISSUE TITLE
[java] Replace Synchronized with ReentrantLock

### ISSUE THREAD

**Issue by numeralnathan** [pmd/pmd#4652]
**Proposed Rule Name:**  ReplaceSynchronized

**Proposed Category:** Multithreading

**Description:**

Java 21 introduces Virtual Threads as an official feature.  Virtual Threads run on a limited number of Carrier Threads.  "synchronized" pins a Virtual Thread to a Carrier Thread.  This reduces the scalability of Virtual Threads and requires manual tuning of the number of Carrier Threads.

Please flag any synchronized and tell the programmer to switch to ReentrantLock.  Please provide a code example on how to do this.

**Code Sample:**

Here is the problematic code.

```java
synchronized (this)   // The PMD rule should flag this as a scalability problem.
{
   // protected block
}
```

Here is the fixed code.  This code creates a new final member field in the called `m_accessLock`.  `m_accessLock` is assigned `new ReentrantLock()` as part of instance initialization.

```java
m_accessLock.lock();

try
{
   // protected block
}
finally
{
   m_accessLock.unlock();
}
```

**Possible Properties:**

I can't think of any properties for this rule.  It should flag violations for synchronized on this, method, class, and parameters.  All of these cases are problematic.

**Comment by zbynek** [comment]
Fixed in #5175
